### ISSUE TITLE
[pyenergyplus] allow running EnergyPlus from arbitrary location

### ISSUE THREAD

**Issue by antoine-galataud** [NREL/EnergyPlus#8482]
Issue overview
--------------

When `api.runtime.run_energyplus()` is called, the expected location of the caller script is the EnergyPlus installation directory. If you consider a complex python app with many modules, or a distributed application where more than 1 python process runs EnergyPlus, it would be much easier to either leave everything in its place or copy relevant EnergyPlus binaries and libraries to an ephemeral location. In both cases, developer should have the possibility to choose the location of EnergyPlus binary when calling the api.
  
Linked with #8433 

### Details

 - Platform: Ubuntu 20.04 LTS
 - Version of EnergyPlus: v9.4.0

**Comment by Myoldmopar** [comment]
@antoine-galataud thanks for your usage of the new API!  I don't actually expect the API calling script to have to live in the EnergyPlus install.  My expectation is that the user will simply ensure that the `pyenergyplus` package can be found in Python's search path.  I usually do this by injecting the preferred EnergyPlus install directory into path prior to importing `pyenergyplus`.  This way I can easily switch between EnergyPlus installs/builds.  My scripts usually have some form of this:

```
sys.path.insert(0, '/path/to/build/or/install/directory')
from pyenergyplus.api import EnergyPlusAPI
```

Would that address your issue?

**Comment by antoine-galataud** [comment]
@Myoldmopar thanks for your answer. I had found this trick in other issues, I should have mentioned it. This partially solves my problem, however on a more complex app structure with parallel runs of EnergyPlus, I need to run each of them from a separate location. Using `sys.path` to do this will be hard to manage. But I totally understand that it's an edge scenario.

From a general perspective, I would find it more flexible and cleaner not to have to rely on `sys.path` to do that, but rather on an api method argument. This would clearly isolate execution contexts.

**Comment by jmarrec** [comment]
@antoine-galataud where would you find the api method in question if you can't load the api to begin with though?

**Comment by jmarrec** [comment]
How about we actually make it a pypi package (similar to what I've done for openstudio)?

I've done a quick test for Ubuntu only, you can try `pip install -i https://test.pypi.org/simple/ pyenergyplus==9.5.0a1`
**The wheel is quite small, just 20MB**, and includes all files required to run energyplus (that is the .py files, the libenergypluslib.so and the new ssc.so), so you're not even required to install energyplus separately.

https://test.pypi.org/project/pyenergyplus/

I did that quickly by adapting a `setup.py` I had created for openstudio, copying the `*.py` and the `libenergyplusapi.so` and `ssc.so`, and using patchelf to set the RPATH of the `libenergyplusapi.so` so that it'd find the `ssc.so` in the current directory instead of inside my build directory... (that was faster for a simple test than getting CMake to do it for me...)

I made one change to `api.py` so that it loads the DLL from the same folder as this. Then used the `setup.py` to build a `bdist_wheel` and used `twine` to upload, and voilà! A python package in under 30 minutes!

```
(py38)julien@~$ ipython
Python 3.8.3 (default, Jun  2 2020, 15:25:16) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.19.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import shlex
   ...: from pyenergyplus import api
   ...: api.api_path()
Out[1]: '/home/julien/Virtualenvs/py38/lib/python3.8/site-packages/pyenergyplus/libenergyplusapi.so'

In [2]: from pyenergyplus.api import EnergyPlusAPI
   ...: api = EnergyPlusAPI()
   ...: state = api.state_manager.new_state()
   ...: return_code = api.runtime.run_energyplus(
   ...:         state,
   ...:         shlex.split('-w /home/julien/Software/Others/EnergyPlus/weather/USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw -d out-ori/ /home/julien/Software/Others/EnergyPlus/testfiles/SolarCollectorFla
   ...: tPlateWater.idf')
   ...: )
EnergyPlus Starting
EnergyPlus, Version 9.5.0-7ac466d702, YMD=2021.02.18 10:28
Adjusting Air System Sizing
Adjusting Standard 62.1 Ventilation Sizing
Initializing Simulation
Reporting Surfaces
Beginning Primary Simulation
Initializing New Environment Parameters
Warming up {1}
Starting Simulation at 01/21 for CHICAGO ANN HTG 99% CONDNS DB
Initializing New Environment Parameters
Warming up {1}
Starting Simulation at 07/21 for CHICAGO ANN CLG 1% CONDNS DB=>MWB
Writing tabular output file results using HTML format.
EnergyPlus Run Time=00hr 00min  0.29sec
EnergyPlus Completed Successfully.
```

**Comment by jmarrec** [comment]
A test to run to see if it works..

```
from pyenergyplus import api
print("api_path={}".format(api.api_path()))
import os
os.system('wget https://github.com/NREL/EnergyPlus/raw/develop/testfiles/1ZoneUncontrolled.idf')
from pyenergyplus.api import EnergyPlusAPI
api = EnergyPlusAPI()
state = api.state_manager.new_state()
return_code = api.runtime.run_energyplus(state,['-D', '-d', 'out/',
                                                '1ZoneUncontrolled.idf'])
```

Note that I had "LINK_WITH_PYTHON=ON" (just learned I could disable it for this purpose), so the wheel I built is linked to libpython3.8.so and will therefore only work on python 3.8...

**Comment by antoine-galataud** [comment]
@jmarrec really cool, I also did something similar on my side for testing.

I believe having the python api bundled separately will help in several regards: packaging of an app that relies on it (you still need EnergyPlus code installed somewhere but it doesn't need to happen at the same time, which makes a difference in build pipelines), and testing (essentially mocked, when core isn't available).

My request on not having to rely on `sys.path` was aiming at this sort of discussion, so I'm glad. However, I'm still convinced that allowing to provide EnergyPlus home installation as a method argument would help, at least for long-lived, distributed executions (whenever it's not only one simulation, not necessarily on the same server, and not sequentially). To make it clearer: it's easier to find where it's installed and provide the path then forget, than managing path over time when application is long-lived. Probably too edgy, I agree.
For the same kind of purpose, I actually already rely on packing and moving necessary libs and binaries around, at it's simpler to manage parallel executions in isolated contexts.

**Comment by Myoldmopar** [comment]
I think the original problem is actually solved now with #8918
